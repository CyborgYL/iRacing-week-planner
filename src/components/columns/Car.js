// @flow

import * as React from 'react';
import classnames from 'classnames';
import intersection from 'lodash.intersection';
import StarIcon from '../icon/StarIcon';
import CarModal from '../modal/CarModal';

import styles from '../../styles/main.module.scss';
import columnStyles from './styles/columns.module.scss';

type Props = {
  race: {
    carIds: Array<number>,
    carClasses: Array<string>,
    series: string,
  },
  favouriteCars: Array<number>,
  ownedCars: Array<number>,
};

export default function Car({ race, favouriteCars, ownedCars }: Props) {
  const [modalOpen, setModalOpen] = React.useState(false);

  const openModal = () => {
    setModalOpen(true);
  };

  return (
    <td
      className={classnames({
        [styles.success]: intersection(ownedCars, race.carIds).length !== 0,
        [columnStyles.clickableCell]: true,
      })}
    >
      <button type="button" className={columnStyles.cellButton} onClick={openModal} onKeyPress={openModal}>
        {intersection(favouriteCars, race.carIds).length !== 0 ? (
          <StarIcon />
        ) : null}
        <span> </span>
        {race.carClasses.join(', ')}
      </button>
      <CarModal
        isOpen={modalOpen}
        onClose={() => setModalOpen(false)}
        ownedCars={ownedCars}
        favouriteCars={favouriteCars}
        carIds={race.carIds}
        seriesName={race.series}
      />
    </td>
  );
}
