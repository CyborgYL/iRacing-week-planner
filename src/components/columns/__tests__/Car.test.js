import React from 'react';
import { shallow } from 'enzyme';
import { describe, test } from '@jest/globals';
import CarModal from '../../modal/CarModal';

import Car from '../Car';

describe('components/columns/Car', () => {
  test('renders correctly', () => {
    const component = shallow(<Car
      race={{
        carIds: [4, 5],
        carClasses: ['Mazda Cup'],
        series: 'Mazda Cup',
      }}
      favouriteCars={[]}
      ownedCars={[]}
    />);

    expect(component).toMatchSnapshot();
  });

  test('renders owned, favourite', () => {
    const component = shallow(<Car
      race={{
        carIds: [4, 5],
        carClasses: ['Mazda Cup'],
        series: 'Mazda Cup',
      }}
      favouriteCars={[4]}
      ownedCars={[5]}
    />);

    expect(component).toMatchSnapshot();
  });

  test('opens modal', () => {
    const component = shallow(<Car
      race={{
        carIds: [4, 5],
        carClasses: ['Mazda Cup'],
        series: 'Mazda Cup',
      }}
      favouriteCars={[4]}
      ownedCars={[5]}
    />);

    component.first().simulate('click');

    expect(component).toMatchSnapshot();

    component.find(CarModal).first().props().onClose();

    expect(component).toMatchSnapshot();
  });
});
