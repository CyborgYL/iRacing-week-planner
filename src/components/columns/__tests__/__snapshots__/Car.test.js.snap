// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/columns/Car opens modal 1`] = `
<td
  className="success clickableCell"
  onClick={[Function]}
>
  <div>
    <StarIcon />
    <span>
       
    </span>
    Mazda Cup
  </div>
  <CarModal
    carIds={
      Array [
        4,
        5,
      ]
    }
    favouriteCars={
      Array [
        4,
      ]
    }
    isOpen={true}
    onClose={[Function]}
    ownedCars={
      Array [
        5,
      ]
    }
    seriesName="Mazda Cup"
  />
</td>
`;

exports[`components/columns/Car opens modal 2`] = `
<td
  className="success clickableCell"
  onClick={[Function]}
>
  <div>
    <StarIcon />
    <span>
       
    </span>
    Mazda Cup
  </div>
  <CarModal
    carIds={
      Array [
        4,
        5,
      ]
    }
    favouriteCars={
      Array [
        4,
      ]
    }
    isOpen={false}
    onClose={[Function]}
    ownedCars={
      Array [
        5,
      ]
    }
    seriesName="Mazda Cup"
  />
</td>
`;

exports[`components/columns/Car renders correctly 1`] = `
<td
  className="clickableCell"
  onClick={[Function]}
>
  <div>
    <span>
       
    </span>
    Mazda Cup
  </div>
  <CarModal
    carIds={
      Array [
        4,
        5,
      ]
    }
    favouriteCars={Array []}
    isOpen={false}
    onClose={[Function]}
    ownedCars={Array []}
    seriesName="Mazda Cup"
  />
</td>
`;

exports[`components/columns/Car renders owned, favourite 1`] = `
<td
  className="success clickableCell"
  onClick={[Function]}
>
  <div>
    <StarIcon />
    <span>
       
    </span>
    Mazda Cup
  </div>
  <CarModal
    carIds={
      Array [
        4,
        5,
      ]
    }
    favouriteCars={
      Array [
        4,
      ]
    }
    isOpen={false}
    onClose={[Function]}
    ownedCars={
      Array [
        5,
      ]
    }
    seriesName="Mazda Cup"
  />
</td>
`;
